### PR TITLE
`ruff server`: Tracing system now respects log level and trace level, with options to log to a file

### LINKED ISSUE #10968: Server log messages should use `window/logMessage`
Currently, the server uses the `tracing` module for log messages which sends everything to stderr. This gets picked up by the client and is added to the log messages at different level regardless of the level used in the server. This level is set by the client.

Here's what VS Code shows:

```
2024-04-16 12:00:50.229 [info] ┐ruff_server::server::api::notifications::did_change::run{file=file:///Users/dhruv/playground/python/animation.py}
┘
2024-04-16 11:59:47.905 [info]    0.101369s ERROR ruff_server::session The following error occurred when trying to find a configuration file at `/Users/dhruv/playground/python`:
No pyproject.toml/ruff.toml/.ruff.toml file was found
   0.101425s ERROR ruff_server::session Falling back to default configuration for `/Users/dhruv/playground/python`
```

Here's what Neovim shows:

```
[START][2024-04-16 12:06:02] LSP logging initiated
[ERROR][2024-04-16 12:06:02] .../vim/lsp/rpc.lua:789	"rpc"	"/Users/dhruv/work/astral/ruff-test/target/debug/ruff"	"stderr"	"   0.001027s DEBUG ruff_server::server No workspace(s) were provided during initialization. Using the current working directory as a default workspace...\n"
[ERROR][2024-04-16 12:06:02] .../vim/lsp/rpc.lua:789	"rpc"	"/Users/dhruv/work/astral/ruff-test/target/debug/ruff"	"stderr"	"   0.041356s ERROR ruff_server::session The following error occurred when trying to find a configuration file at `/Users/dhruv/playground/python`:\nNo pyproject.toml/ruff.toml/.ruff.toml file was found\n   0.041386s ERROR ruff_server::session Falling back to default configuration for `/Users/dhruv/playground/python`\n"
[ERROR][2024-04-16 12:06:02] .../vim/lsp/rpc.lua:789	"rpc"	"/Users/dhruv/work/astral/ruff-test/target/debug/ruff"	"stderr"	"   0.044606s WARN ruff_server::server LSP client does not support dynamic capability registration - automatic configuration reloading will not be available.\n┐ruff_server::server::api::notifications::did_open::run{file=file:///Users/dhruv/playground/python/errors.py}\n┘\n"
```

This is not a high priority. We would need to look at all the messages using `tracing` and convert most of them to use the notification request instead.

### LINKED ISSUE #11545: Ruff server not closing with neovim instances - vol 2
I am still seeing this problem originally reported (and fixed) in #11207 
I am on ruff 0.4.5 and neovim 0.9.5 on Linux Fedora 39.

Here's the output from Lsp logs

```
[START][2024-05-26 16:20:07] LSP logging initiated
  16021 [ERROR][2024-05-26 16:20:07] .../vim/lsp/rpc.lua:734»   "rpc"»  "/home/hollas/.local/share/nvim/mason/bin/ruff"»"stderr"»       "wa        rning: The top-level linter settings are deprecated in favour of their counterparts in the `lint` section. Please update the follow        ing options in `.venv/lib64/python3.12/site-packages/pandas/pyproject.toml`:\n  - 'ignore' -> 'lint.ignore'\n  - 'select' -> 'lint.        select'\n  - 'typing-modules' -> 'lint.typing-modules'\n  - 'unfixable' -> 'lint.unfixable'\n  - 'per-file-ignores' -> 'lint.per-fi        le-ignores'\n"
  16022 [ERROR][2024-05-26 16:20:07] .../vim/lsp/rpc.lua:734»   "rpc"»  "/home/hollas/.local/share/nvim/mason/bin/ruff"»"stderr"»       "wa        rning: `PGH001` has been remapped to `S307`.\n"
  16023 [ERROR][2024-05-26 16:20:07] .../vim/lsp/rpc.lua:734»   "rpc"»  "/home/hollas/.local/share/nvim/mason/bin/ruff"»"stderr"»       "           0.700874s WARN ruff_server::server LSP client does not support dynamic capability registration - automatic configuration reloading         will not be available.\n"
  16024 [ERROR][2024-05-26 16:20:16] .../vim/lsp/rpc.lua:734»   "rpc"»  "/home/hollas/.local/share/nvim/mason/bin/ruff"»"stderr"»       "           9.789324s INFO ruff_server::server::connection Shutdown request received. Waiting for an exit notification...\n"
```

_Originally posted by @danielhollas in https://github.com/astral-sh/ruff/issues/11207#issuecomment-2128365016_

cc @dhruvmanila

### PR DESCRIPTION
**Pull request by snowsignal**
## Summary

Fixes #10968.
Fixes #11545.

The server's tracing system has been rewritten from the ground up. The server now has trace level and log level settings which restrict the tracing events and spans that get logged. 

* A `logLevel` setting has been added, which lets a user set the log level. By default, it is set to `"info"`.
* A `logFile` setting has also been added, which lets the user supply an optional file to send tracing output (it does not have to exist as a file yet). By default, if this is unset, tracing output will be sent to `stderr`.
* A `$/setTrace` handler has also been added, and we also set the trace level from the initialization options. For editors without direct support for tracing, the environment variable `RUFF_TRACE` can override the trace level.
* Small changes have been made to how we display tracing output. We no longer use `tracing-tree`, and instead use `tracing_subscriber::fmt::Layer` to format output. Thread names are now included in traces, and I've made some adjustment to thread worker names to be more useful.

## Test Plan

In VS Code, with `ruff.trace.server` set to its default value, no logs from Ruff should appear.

After changing `ruff.trace.server` to either `messages` or `verbose`, you should see log messages at `info` level or higher appear in Ruff's output:
Screenshot 2024-06-10 at 10 35 04 AM

In Helix, by default, no logs from Ruff should appear.

To set the trace level in Helix, you'll need to modify your language configuration as follows:
```toml
[language-server.ruff]
command = "/Users/jane/astral/ruff/target/debug/ruff"
args = ["server", "--preview"]
environment = { "RUFF_TRACE" = "messages" }
```

After doing this, logs of `info` level or higher should be visible in Helix:
Screenshot 2024-06-10 at 10 39 26 AM

You can use `:log-open` to quickly open the Helix log file.

In Neovim, by default, no logs from Ruff should appear.

To set the trace level in Neovim, you'll need to modify your configuration as follows:
```lua
require('lspconfig').ruff.setup {
  cmd = {"/path/to/debug/executable", "server", "--preview"},
  cmd_env = { RUFF_TRACE = "messages" }
}
```

You should see logs appear in `:LspLog` that look like the following:
Screenshot 2024-06-11 at 11 24 01 AM

You can adjust `logLevel` and `logFile` in `settings`:
```lua
require('lspconfig').ruff.setup {
  cmd = {"/path/to/debug/executable", "server", "--preview"},
  cmd_env = { RUFF_TRACE = "messages" },
  settings = {
    logLevel = "debug",
    logFile = "your/log/file/path/log.txt"
  }
}
```

The `logLevel` and `logFile` can also be set in Helix like so:
```toml
[language-server.ruff.config.settings]
logLevel = "debug"
logFile = "your/log/file/path/log.txt"
```

Even if this log file does not exist, it should now be created and written to after running the server:

Screenshot 2024-06-10 at 10 43 44 AM

### CODE REVIEW

File: crates/ruff_server/src/trace.rs
Comment (snowsignal) [inline crates/ruff_server/src/trace.rs:11]: The reason I went with `Messages` instead of `Off` as a default is so that `info` events are logged by default if the trace value is unset by the LSP client.

**Review by MichaReiser — CHANGES_REQUESTED** [review]
Looks good. I think we should address the TODO before merging this and extend the test plan with a case that shows that changing the verbosity in the extension settings changes the output.

File: crates/ruff_server/src/trace.rs
Comment (snowsignal) [reply crates/ruff_server/src/trace.rs:18]: The PR made a note that this would be used in the future:

> This PR also introduces a $/setTrace handler, which will be used in the future by our tracing span subscriber.

I wanted to at least add a handler for this so that we don't get a warning about a nonexistent handler every time the trace level changes.

File: crates/ruff_server/src/trace.rs
Comment (snowsignal) [reply crates/ruff_server/src/trace.rs:25]: I actually explored having the log level be set by the trace value (you can see it in the commit history). I ultimately decided against this because the trace value sent by the LSP is [only meant for `$/logTrace`, not `window/logMessage`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#traceValue).

Comment (snowsignal) [reply crates/ruff_server/src/trace.rs]: I've introduced a `logLevel` setting to client settings to fix this.

Comment (MichaReiser) [reply crates/ruff_server/src/trace.rs]: Oh I didn't realize that difference when reviewing yesterday. What's the reason we don't use `$/logTrace` because it seems like the right fit from the spec

> $/logTrace should be used for systematic trace reporting. For single debugging messages, the server should send [window/logMessage](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_logMessage) notifications. [source](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#logTrace)

Comment (MichaReiser) [reply crates/ruff_server/src/trace.rs]: Okay, I think I'm starting to understand what this PR does. I think the PR title needs updating. 

I try to summarize my understanding to make sure it is correct. 

This PR only logs `trace::debug`, `trace::warn` but it omits all `trace::debug_span` etc. Is this correct?

What's the motivation for making this distinction instead of just logging all tracing events with `$/logTrace` for now. That seems simpler because it doesn't require additional filtering or yet another setting (which I'm worried that it just gets hard to understand for users, especially because `TRACE_VALUE: message` exists). Is it because you need a way to format a span to a string?

Comment (snowsignal) [reply crates/ruff_server/src/trace.rs]: @MichaReiser My plan was for tracing events to go through `window/logMessage` and then later implement a additional tracing layer that sends all tracing events / messages to `$/logTrace` as a follow-up.

The reason I'm doing this is to separate the log level from the trace level.

File: crates/ruff_server/src/trace.rs
Comment (MichaReiser) [inline crates/ruff_server/src/trace.rs:131]: Nit: `tracing-subscriber` has a `BoxMakeWriter` that is just that. You can call it with `BoxMakeWriter::new(std::io::stderr)`

File: crates/ruff_server/src/trace.rs
Comment (MichaReiser) [inline crates/ruff_server/src/trace.rs:143]: I would find some module level documentation helpful that explains the different ways on how tracing can be configured. It seem we now support an env var, `TRACE_VALUE`, a file, and the PR description mentions a `logLevel` setting, but it's not clear to me where that one is set.

File: crates/ruff_server/src/trace.rs
Comment (MichaReiser) [inline crates/ruff_server/src/trace.rs:103]: It seems that we now only support `Off` vs `Message` and `Verbose`. What do you think of mapping `Message` to `Info`, and `Verbose` to `Debug` or `Trace`?

File: crates/ruff_server/src/trace.rs
Comment (MichaReiser) [inline crates/ruff_server/src/trace.rs:95]: It seems we need to inform tracing  when the log level changes

> If the maximum level the Filter will enable can change over the course of its lifetime, it is free to return a different value from multiple invocations of this method. However, note that changes in the maximum level will only be reflected after the callsite Interest cache is rebuilt, by calling the tracing_core::callsite::rebuild_interest_cache function. Therefore, if the Filter will change the value returned by this method, it is responsible for ensuring that [rebuild_interest_cache`]rebuild is called after the value of the max level changes.

[source](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Filter.html#method.max_level_hint)

File: crates/ruff_server/src/trace.rs
Comment (MichaReiser) [inline crates/ruff_server/src/trace.rs:76]: What's the motivation of having two filters rather than just one?

File: crates/ruff_server/src/message.rs
Comment (MichaReiser) [inline crates/ruff_server/src/message.rs:38]: Did you change this to `eprintln` because it might happen before tracing is set up? I think it would still be nice if the message also goes to tracing to ensure the error is captured in a logfile if `logFile` is set

File: crates/ruff_server/src/session/settings.rs
Comment (MichaReiser) [inline crates/ruff_server/src/session/settings.rs:77]: How about creating a `GlobalSettings` struct that includes a `client_settings` fields attributed with `#[flatten]`?

File: crates/ruff_server/src/server.rs
Comment (MichaReiser) [inline crates/ruff_server/src/server.rs:87]: Is it required to restart the server when changing the `logFile` or `logLevel` setting?

**Review by MichaReiser — COMMENTED** [review]
The code changes look good to me. I left a few questions and are waiting for a test plan that shows a few traces. 

I think some documentation on how the tracing setup would be helpful, especially because it now supports environment variables, tracing level, logLevel, log file etc.

File: crates/ruff_server/src/trace.rs
Comment (snowsignal) [reply crates/ruff_server/src/trace.rs:103]: The difference between `messages` and `verbose` is the amount of information that gets logged, not the level of logging. Since our tracing system doesn't have separated 'verbose' information attached to its traces, I think it makes sense to treat them the same. `logLevel` should be the source of truth for the traces that get shown, IMO.

File: crates/ruff_server/src/trace.rs
Comment (snowsignal) [reply crates/ruff_server/src/trace.rs:95]: The log level doesn't change over the lifecycle of the server, though. It's the responsibility of the client to reboot the server with new settings when the client settings change.

File: crates/ruff_server/src/trace.rs
Comment (snowsignal) [reply crates/ruff_server/src/trace.rs:76]: They're filtering by different things, and I'd rather keep them separated for modularity.

File: crates/ruff_server/src/message.rs
Comment (snowsignal) [reply crates/ruff_server/src/message.rs:38]: Good idea!

File: crates/ruff_server/src/server.rs
Comment (snowsignal) [reply crates/ruff_server/src/server.rs:87]: Yes, that's how its set up right now.

File: crates/ruff_server/src/trace.rs
Comment (MichaReiser) [reply crates/ruff_server/src/trace.rs:76]: Could you add a documentation explaining what they're filtering by then so that a future me understand the motivation of keeping them different and doesn't merge them together.

File: crates/ruff_server/src/trace.rs
Comment (snowsignal) [reply crates/ruff_server/src/trace.rs:131]: Unfortunately, this is a different return type, so it wouldn't work here. We need to consolidate the log file / stderr writer into a single type here, otherwise the tracing layer would have a different type and we'd need to build it in two completely distinct branches.

Comment (MichaReiser) [reply crates/ruff_server/src/trace.rs:131]: It seems to work fine for rust analyzer:

```
    let writer = match log_file {
        Some(file) => BoxMakeWriter::new(Arc::new(file)),
        None => BoxMakeWriter::new(std::io::stderr),
    };
```

**Comment by MichaReiser** [comment]
This looks good to me. Happy to approve tomorrow morning when the Test plan's filled out. Thanks for doing another iteration on this!

File: crates/ruff_server/src/trace.rs
Comment (snowsignal) [reply crates/ruff_server/src/trace.rs:149]: Oh, I see what you mean! Nevermind 😄

**Review by MichaReiser — APPROVED** [review]
Nice! Thank's for taking the time to iterate on the design

File: crates/ruff_server/src/trace.rs
Comment (dhruvmanila) [inline crates/ruff_server/src/trace.rs:16]: It might be useful to provide this information in the server docs either in the global docs (`crates/ruff_server/README.md`) or editor specific docs (`crates/ruff_server/docs/setup/*.md`).

Comment (snowsignal) [reply crates/ruff_server/src/trace.rs:16]: I'll update the docs in a follow-up.

**Comment by snowsignal** [comment]
@dhruvmanila I've updated the PR description with a Neovim testing guide 😄

**Pull request merged**
